java/flight/src/main/java/org/apache/arrow/flight/FlightService.java

```diff
@@ -175,7 +185,14 @@ public void putNext() {
     @Override
     public void putNext(ArrowBuf metadata) {
       Preconditions.checkNotNull(unloader);
-      responseObserver.onNext(new ArrowMessage(unloader.getRecordBatch(), metadata));
+      // close is a no-op if the message has been written to gRPC, otherwise frees the associated buffers
+      // in some code paths (e.g. if the call is cancelled), gRPC does not write the message, so we need to clean up
+      // ourselves. Normally, writing the ArrowMessage will transfer ownership of the data to gRPC/Netty.
+      try (final ArrowMessage message = new ArrowMessage(unloader.getRecordBatch(), metadata)) {
+        responseObserver.onNext(message);
+      } catch (Exception e) {
+        throw new RuntimeException("Error freeing ArrowMessage", e);
```

Comment:
I would typically log this as opposed to throw since I'm not sure what you're going to be able to do about it. Who exactly does this throw to?

Reply:
This would have gotten thrown to the client, but I changed ArrowMessage to not throw a checked exception instead (since all its close method does is free buffers).

Reply:
I changed this back again, to throw a runtime exception if `ArrowMessage#close` does throw. My reasoning is as follows: generally, `ArrowMessage#close` should not throw. But if it does, although the user can't reasonably do anything about this, we shouldn't let the application continue since we don't know what other state might be in an odd state (and presumably, the only reason for something to throw that doesn't usually throw is because some state got corrupted somewhere).